llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -2859,6 +2861,32 @@ static SDValue lowerCTLZ_CTTZ_ZERO_UNDEF(SDValue Op, SelectionDAG &DAG) {
   return DAG.getNode(ISD::SUB, DL, VT, DAG.getConstant(Adjust, DL, VT), Trunc);
 }
 
+/// The CToPtr instruction is deprecated and will be removed. This function
+/// emits the currently defined semantics of `gettag(x) ? x.addr - base : 0`.
+static SDValue emitCToPtrReplacement(SelectionDAG &DAG, const SDLoc &DL,
+                                     SDValue Cap, SDValue Base, EVT XLenVT) {
+  // Instead of emitting a select instruction (which will be lowered to a branch
+  // since RISC-V lacks conditional move), we can use cgetaddr(x) & -cgettag(x).
+  // This ensures that an unset tag results in an all zeroes mask and a valid
+  // tag bit gives use -1, i.e. all ones.
+  // Note: With Zicond we could lower this to czero.eqz instead.
+  // Ideally we would keep the select here and allow follow-up folds to optimize
+  // the select depending on available instructions, but the MVT::i1 intrinsic
+  // call is optimized poorly, so we expand it manually.
+  SDValue IsTagged = DAG.getNode(RISCVISD::CAP_TAG_GET, DL, XLenVT, Cap);
+  SDValue Mask = DAG.getNode(ISD::SUB, DL, XLenVT,
+                             DAG.getConstant(0, DL, XLenVT), IsTagged);
+  // Using EXTRACT_SUBREG instead of getaddr is safe here since the result is
+  // only used by the AND node, so no capability metadata is leaked.
+  SDValue Addr =
+      DAG.getTargetExtractSubreg(RISCV::sub_cap_addr, DL, XLenVT, Cap);
+  // NB: For the ISAv9 we no longer have DDC relocation, so when expanding
+  // PTRTOINT we do not subtract a base value.
+  if (Base)
+    Addr = DAG.getNode(ISD::SUB, DL, XLenVT, Addr, Base);
```

Comment:
This makes `(ptrtoint foo)` behave differently to `(cfromptr foo, (cspecialr ddc))`. Shouldn't we be consistent and never subtract the base?

Reply:
That was my initial thought, but I then decided to not change the semantics of the existing instruction. Another option could be to completely remove the builtin for ISAv9 and let people spell out exactly how they want to convert int->cap?

Reply:
I think it has to be consistent if it exists. CheriBSD has:

```
```

Reply:
Also note that Morello, since it uses the CVTP instructions, has it follow the CCTLR config like inttoptr

Reply:
(See cheribsdtest test_nofault_cfromptr)

Reply:
Fixed to always use setaddr.